src/main/ruby/core/truffle/cext.rb

```diff
@@ -1192,11 +1192,9 @@ def rb_str_conv_enc_opts(str, from, to, ecflags, ecopts)
       end
       return str
     end
-    begin
-      rb_str_encode(str, to, ecflags, ecopts)
-    rescue Encoding::InvalidByteSequenceError
-      str
-    end
+
+    ec = Encoding::Converter.new(from, to, ecopts || ecflags)
+    ec.convert(str)
```

Comment:
There may be some small difference between `ec.convert` and `rb_str_conv_enc_opts` for example handling infecting the result and returning the original string if conversion fails?

Reply:
I'll take another look with an eye out for that. Thanks.

Reply:
Taint state is preserved. But the previous code didn't always return the string if the encoding failed. It wouldn't catch `Encoding::UndefinedConversionError`, which is the error type I kept seeing.

If you want to avoid `Encoding::Converter`, we could skip the `rb_str_encode` call and call `String#encode` with the the from & to encodings. That should go through the same path, just a bit longer.

Reply:
I'll leave it up to you to decide I just wanted to point out the potential incompatibility, thanks.

Reply:
You were right about the failure case. I updated using `primitive_convert` to use a non-exceptional method that reports status we can branch on.